core/src/main/scala/laserdisc/protocol/RESP.scala

```diff
@@ -350,13 +376,67 @@ trait RESPCodecs { this: RESPBuilders =>
           case `colon`  => Right(integerCodec)
           case `dollar` => Right(bulkStringCodec)
           case `star`   => Right(arrayCodec)
-          case other    => Left(s"unidentified RESP type ${other.toHex}")
+          case other    => Left(s"unidentified RESP type (Hex: ${other.toHex})")
         }
         .fold(
-          error => Attempt.failure(Err(error)), { case (remainder, codec) => codec.decode(remainder) }
+          error => Attempt.failure(Err(error)),
+          { case (remainder, codec) => codec.decode(remainder) }
         )
+
     override final def toString: String = "RESP"
   }
 }
 
-object RESP extends RESPBuilders with RESPCodecs
+sealed trait RESPFunctions { this: RESPCodecs =>
+
+  import BitVectorDecoding._
+
+  final def stateOf: BitVector => String | BitVectorState =
+    bits => bits.consume(BitsInByte) {
+      case `plus`
+           | `minus`
+           | `colon` => Right(No)
+      case `dollar`  => Right(Bulk)
+      case `star`    => Right(Multi)
+      case other     => Left(s"unidentified RESP type (Hex: ${other.toHex})")
+    }
+    .flatMap {
+      case (remainder, Bulk) =>
+        evalWithSizeDecodedFrom(remainder) {
+          ds =>
+            if (ds.value >= 0 && ds.remainder.size == (crlf.size + (ds.value * BitsInByte))) CompleteVector // Complete: expected size
+            else if (ds.value == -1) CompleteVector // Complete: empty bulk
+            else MissingBits((crlf.size + (ds.value * BitsInByte)) - ds.remainder.size)
+        }
+
+      case (remainder, Multi) =>
+        arrayCodec.decode(remainder).fold(
```

Comment:
pity, but I don't blame you 😛

Reply:
yes, that's a shortcut. I might try the full travers / recurse /traverse / recurse ... later. But as discussed offline I'm afraid it could be slower than this anyway.